serverless.yaml

```diff
@@ -81,16 +81,20 @@ functions:
           method: ANY
           path: /
           private: true
+          cors: true
       - http:
           method: ANY
           path: '{proxy+}'
           private: true
+          cors: true

```

Comment:
Both level need CORS enabled. The browser will send out pre-flight requests (OPTION method) to API Gateway, and API Gateway will return CORS headers. And then the browser will send out real API requests (GET method) to API Gateway, and fhir-server lambda function  (CORS middleware) will return CORS headers.

I know it looks like pre-flight is not meaningful, if the browser requires CORS headers from real API calls. But it is how browser works.

Reply:
You can find more details about CORS from this document https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

Reply:
![Capture](https://user-images.githubusercontent.com/88723283/129201571-89d5e4b7-f83f-4f96-993d-8be798830d97.PNG)

Reply:
Does the CORS middleware not handle preflight OPTIONS requests?

Reply:
Yes, the CORS middleware handles preflights requests. I have removed the cors in serverless.yml

Reply:
Awesome.  Can the /metadata and /smart-configuration endpoints also have CORS removed at the API Gateway level just like we did here for /{proxy+}?  Or are they different?

Reply:
I removed CORS for /metadata and /smart-configuration at the API Gateway level. I had to add CORS for /{proxy+} at API Gateway level. Because /{proxy+} is private, without the API key, OPTION method request issued by the browser cannot get to Lambda function. In a word, any private path needs API Gateway level CORS support

Reply:
> I had to add CORS for /{proxy+} at API Gateway level. Because /{proxy+} is private, without the API key, OPTION method request issued by the browser cannot get to Lambda function. 

Sorry, I'm having a difficult time understand that.  So right now, I'm looking at the API Gateway /{proxy+} resource.  It has an _ANY_ HTTP method that seems to be setup to forward/proxy all request to the integrating Lambda.  It also has an OPTIONS method that you put there.  Are you saying that if the OPTIONS method was removed from the /{proxy+} resource, API Gateway wouldn't forward the request to the Lambda to handle it?

Reply:
API Gateway wouldn't forward OPTIONS request to Lambda to handle it.

Reply:
OK, that makes sense.  Thanks for explaining.  

One more question.  When I select the OPTIONS method on the /{proxy+} resource, I see it's integrated with a _Mock Endpoint_.  The other OPTIONS methods on /metadata and /smart-configuration are integrated with the Lambda.  What is the reason for having the OPTIONS method is integrated with _Mock Endpoint_ and not the Lambda?

Reply:
If OPTIONS method for /{proxy+} resource is integrated with the Lambda, then API Gateway would return HTTP 403 for OPTIONS requests from the browser. Because /{proxy+} resource is private, and it requires API key in header of OPTIONS requests. If OPTIONS method is integrated with Mock Endpoint, then API Gateway doesn't require API key header, and returns CORS headers.

Reply:
OK, just to record here what we talked about in IM.
the OPTIONS method on the /{proxy+} resource wasn't marked as `private: true` in serverless.yml, and so _didn't_ need an API key header.  Because of this, it was possible to integrate the OPTIONS method with the Lambda instead of integrating it with a Mock Endpoint.  Also, the serverless.yaml `cors: true` was also removed on the `/{proxy+}` resource because the Lambda is now handling all theings CORS.  Did I get that right Linden?

Reply:
yes